rclpy/src/rclpy/_rclpy.c

```diff
@@ -2834,31 +3141,50 @@ rclpy_get_node_names_and_namespaces(PyObject * Py_UNUSED(self), PyObject * args)
   }
 
   PyObject * pynode_names_and_namespaces = PyList_New(node_names.size);
+  if (!pynode_names_and_namespaces) {
+    goto cleanup;
+  }
   size_t idx;
   for (idx = 0; idx < node_names.size; ++idx) {
     PyObject * pytuple = PyTuple_New(2);
-    PyTuple_SetItem(
-      pytuple, 0,
-      PyUnicode_FromString(node_names.data[idx]));
-    PyTuple_SetItem(
-      pytuple, 1,
-      PyUnicode_FromString(node_namespaces.data[idx]));
-    PyList_SetItem(
-      pynode_names_and_namespaces, idx,
-      pytuple);
-  }
-
-  ret = rcutils_string_array_fini(&node_names);
-  if (ret != RCUTILS_RET_OK) {
+    if (!pytuple) {
+      goto cleanup;
+    }
+    PyObject * pynode_name = PyUnicode_FromString(node_names.data[idx]);
+    if (!pynode_name) {
+      Py_DECREF(pytuple);
+      goto cleanup;
+    }
+    // Steals the reference
+    PyTuple_SET_ITEM(pytuple, 0, pynode_name);
+    PyObject * pynode_namespace = PyUnicode_FromString(node_namespaces.data[idx]);
+    if (!pynode_namespace) {
+      Py_DECREF(pytuple);
+      goto cleanup;
+    }
+    // Steals the reference
+    PyTuple_SET_ITEM(pytuple, 1, pynode_namespace);
+    // Steals the reference
+    PyList_SET_ITEM(pynode_names_and_namespaces, idx, pytuple);
+  }
+
+  rcutils_ret_t fini_names_ret;
+  rcutils_ret_t fini_namespaces_ret;
+cleanup:
+  fini_names_ret = rcutils_string_array_fini(&node_names);
+  fini_namespaces_ret = rcutils_string_array_fini(&node_namespaces);
```

Comment:
@jacobperron nit: can't we declare and initialize on the same statement?

Reply:
Apparently, this is not supported and results in compiler errors. This [SO post](https://stackoverflow.com/a/8384505) suggest we could add an empty statement as a workaround. I don't feel strongly either way, but I'm leaning towards leaving it as-is because `cleanup:;` looks odd to me.

Reply:
I'm actually surprised you don't have warnings above where you declare local variables _after_ the use of a `goto`. In the `rmw` implementations we declare all local variables just before the first use of `goto`, which is essentially what the compiler does even when you declare them throughout and I gather the use of goto might cause strange order of side-effects if you don't declare everything ahead of their use.

For example:

https://github.com/ros2/rmw_connext/blob/43d6833977ddfc3d8fa599a51c7b01bcb3008fa3/rmw_connext_shared_cpp/src/node.cpp#L119-L140

Reply:
Yeah, in retrospect this is odd.
I am able to produce an error if I try to both declare and initialize *after* a goto statement. But the compiler is fine if it's only a declaration. I've moved the declarations to the top so code reads better 67d3660